pkg/app/server.go

```diff
@@ -323,10 +358,101 @@ func RunKubeStateMetrics(ctx context.Context, opts *options.Options) error {
+	defer olderCancel()
+	updateFn := func(prefix string) {
+		klog.Infof("%s: updating stores", prefix)
+		// Generate config for when none of the CRDs were present during callers' initialization.
+		// This is to ensure that the config does not throw an EOF error due to coming up empty.
+		customConfig, err := resolveCustomResourceConfig(opts)
+		if err != nil {
+			klog.Errorf("%s: failed to update custom resource stores", prefix)
+		}
+		// Fetch the discovered factories.
+		gvkMaps.NewGVKFactoriesFn, err = customresourcestate.ResolveConfig(customConfig, gvkMaps)
+		if err != nil {
+			klog.Errorf("%s: failed to update custom resource stores: %v", prefix, err)
+		}
+		// Get families for discovered factories.
+		customFactories, err := gvkMaps.NewGVKFactoriesFn()
+		if err != nil {
+			klog.Errorf("%s: failed to update custom resource stores: %v", prefix, err)
+		}
+		// Update the list of enabled custom resources.
+		for _, factory := range customFactories {
+			enabledCustomResources = append(enabledCustomResources, factory.Name())
+		}
+		// Create clients for discovered factories.
+		_, discoveredCustomResourceClients, err := createKubeClient(opts.Apiserver, opts.Kubeconfig, customFactories...)
+		if err != nil {
+			klog.Errorf("%s: failed to update custom resource stores: %v", prefix, err)
+		}
+		// Update the store builder with the new clients.
+		storeBuilder.WithCustomResourceClients(discoveredCustomResourceClients)
+		// Inject families' constructors to the existing set of stores.
+		storeBuilder.WithCustomResourceStoreFactories(customFactories...)
+		// Update the store builder with the new custom resources.
+		if err := storeBuilder.WithEnabledResources(enabledCustomResources); err != nil {
+			klog.Errorf("%s: failed to update custom resource stores: %v", prefix, err)
+		}
+		// Configure the generation function for the custom resource stores.
+		storeBuilder.WithGenerateCustomResourceStoresFunc(storeBuilder.DefaultGenerateCustomResourceStoresFunc())
+		// Reset the flag, if there were no errors. Else, we'll try again on the next tick.
+		gvkMaps.M.Lock()
+		gvkMaps.UpdateStores = false
+		gvkMaps.M.Unlock()
+		// Run the metrics handler with updated configs.
+		olderCancel()
+		olderContext, olderCancel = context.WithCancel(ctx)
+		err = m.Run(olderContext)
+		if err != nil {
+			klog.Errorf("%s: failed to run metrics handler: %v", prefix, err)
+			t.Stop()
+			return
+		}
+	}
+	go func() {
+		for range t.C {
+			select {
+			case <-ctx.Done():
+				t.Stop()
+				return
+			default:
+				// Prefix to be used for all log messages.
+				prefix := "discovery ticker"
+				// Subsequent discoveries will only occur if a discovery was previously made.
+				klog.Infof("%s: discovering resources", prefix)
+				// TODO: This takes (at least) a good minute, because `ServerGroups` and `ServerResourcesForGroupVersion` are both synchronous.
+				gvkMaps.Discover()
+				if gvkMaps.UpdateStores {
```

Comment:
You need to lock against reads too, if you're locking writes.

Reply:
Ah, my bad. Done! :)